lib/EnsEMBL/REST/Model/ga4gh/Beacon.pm

```diff
@@ -468,48 +594,103 @@ sub variant_exists {
+        my $response = get_dataset_allele_response($dataset_var_found{$dataset_id}, $assemblyId, 1, $vf_found, $error, $sv);
+        push (@dataset_response, $response);
+      }
+    }
+    # ALL - returns all datasets even those that don't have the queried variant
+    # If there is a list of datasets then dataset response returns all of them, if not then returns all available datasets
+    elsif ($incl_ds_response == 1) {
+      %datasets = $has_dataset ? %variation_set_list : %available_datasets;
+      my $found_in_dataset = $vf_found ? 1 : 0;
+      foreach my $dataset_id (keys %datasets) {
+        if (exists $dataset_var_found{$dataset_id}){
+          my $response = get_dataset_allele_response($dataset_var_found{$dataset_id}, $assemblyId, $found_in_dataset, $vf_found, $error, $sv);
+          push (@dataset_response, $response);
+        }
+        else {
+          my $response = get_dataset_allele_response($datasets{$dataset_id}, $assemblyId, 0, $vf_found, $error, $sv);
+          push (@dataset_response, $response);
+        }
+      }
+    }
+    # MISS - means opposite to HIT value, only datasets that don't have the queried variant
+    # Same as HIT but only the datasets that don't have the variant are returned
+    elsif ($incl_ds_response == 3) {
+      %datasets = $has_dataset ? %variation_set_list : %available_datasets;
+      foreach my $dataset_id (keys %datasets) {
+        if (!exists $dataset_var_found{$dataset_id}){
+          my $response = get_dataset_allele_response($datasets{$dataset_id}, $assemblyId, 0, $vf_found, $error, $sv);
+          push (@dataset_response, $response);
+        }
+      }
+    }
   }
-  return ($found, $dataset_response);
+
+  return ($found,\@dataset_response);
 }
 
 # Returns a BeaconDatasetAlleleResponse for a
 # variant feature
 # Assumes that it exists
 sub get_dataset_allele_response {
-  my ($dataset, $found, $vf, $error) = @_;
- 
+  my ($dataset, $assemblyId, $found, $vf, $error, $sv) = @_;
+
   my $ds_response;
+    $ds_response->{'datasetId'} = $dataset->short_name();
+    $ds_response->{'exists'} = undef;
+    $ds_response->{'error'} = undef;
+    $ds_response->{'frequency'} = undef;
+    $ds_response->{'variantCount'} = undef;
+    $ds_response->{'callCount'} = undef;
+    $ds_response->{'sampleCount'} = undef;
+    $ds_response->{'note'} = undef;
+    $ds_response->{'externalUrl'} = undef;
+    $ds_response->{'info'} = undef;
+
+    # Change 
+    if (! defined $found) {
+      $ds_response->{'error'} = $error;
+      return $ds_response;
+    }
+    if ($found == 0) {
+      $ds_response->{'exists'} = JSON::false;
+      return $ds_response;
+    }
+
+    $ds_response->{'exists'} = JSON::true;
+
+    my $externalURL = "http://www.ensembl.org";
+    if ($assemblyId eq 'GRCh37') {
+      $externalURL = "http://grch37.ensembl.org";
+    }
+
+    if($sv == 1){
+      $externalURL .= "/Homo_sapiens/StructuralVariation/Explore?sv=" . $vf->variation_name() unless !defined($vf);
```

Comment:
unless of a negative condition is very confusing and error prone... as the PBP book puts it: "Don’t Fail Not to Use Chained Negatives"
either change it to a if or better yet, add the condition to the if you're doing on the previous line.
It will be a lot more clear that way.

Reply:
Good point. Thanks.